### DISCUSSION TITLE
Is it possible to embed the launching of a Codespace template and dev container from a website?

### DISCUSSION THREAD

**Discussion by eighty4** [community/community#44933] [Codespaces] [answered]
Hi,

I'd like to create a button on my webpage that sends a user to GitHub to launch a Codespace for a repository and dev container. Is is possible to parameterize `https://github.com/codespaces/new` or to embed with a JS API?

Thanks,
Adam

**Comment by jge162** [comment, marked as answer]
Hello @eighty4 , it is possible to parameterize the URL for creating a new Codespace on GitHub. You can use query parameters to pre-populate the repository and container values in the "New Codespace" form.

For example, the following URL would pre-populate the form with the repository "my-repo" and the container "my-container":


‘’’
https://github.com/codespaces/new?repository=my-repo&container=my-container
‘’’

Alternatively, you can use the GitHub Codespaces API to create a Codespace programmatically. This would require authentication and authorization, and you would need to use an access token with the appropriate scopes.

You can embed the GitHub Codespaces API in your webpage using Javascript, for example by using the fetch function to make a POST request to the API endpoint.

You could also use the GitHub OAuth API to get an access token for the user, which you can then use to make API requests on their behalf.

Be aware that creating a codespace may require for the user to have a GitHub Pro or GitHub Enterprise subscription.

**Reply by eighty4** [reply]
Does authorizing with oauth require a server side deployment to handle the auth code flow/redirect or can it all be done in JavaScript? Would a static page with JavaScript be able to get an access token?

**Reply by eighty4** [reply]
Is it possible to dynamically create a devcontainer.json config using the API or does it have to exist as a file in the repository?

**Reply by eighty4** [reply]
Does the URL `https://user-my-cool-codespace-asdfasdgasdg.github.dev/` have similar file/line params as the repository tree linking features of github.com with the `L` line number specified?

**Reply by jge162** [reply]
> Does authorizing with oauth require a server side deployment to handle the auth code flow/redirect or can it all be done in JavaScript? Would a static page with JavaScript be able to get an access token?

A static page with JavaScript can get an access token using the Implicit Grant flow

**Reply by jge162** [reply]
> Is it possible to dynamically create a devcontainer.json config using the API or does it have to exist as a file in the repository?

It is also possible to use the API to read the devcontainer.json config from the API and then use it to configure the container. There are several libraries and tools that allows to do it. but it is not a built in feature of VSC...

**Reply by jge162** [reply]
> Does the URL `https://user-my-cool-codespace-asdfasdgasdg.github.dev/` have similar file/line params as the repository tree linking features of github.com with the `L` line number specified?

GitHub Codespaces do not have the same file/line linking features as the repository tree on github.com, but they do have some similar functionality. However, Codespaces do not have the ability to specify a line number in the URL like you can with the repository tree on github.com.

**Reply by eighty4** [reply]
Thanks for all the answers!

**Comment by aaronsteers** [comment]
Link to related: 

- [URL to directly launch a new codespace?](https://github.com/orgs/community/discussions/39208)
